src/main/AllocationTable.cpp

```diff
@@ -0,0 +1,334 @@
+/*
+ *  R : A Computer Language for Statistical Data Analysis
+ *  Copyright (C) 1995, 1996  Robert Gentleman and Ross Ihaka
+ *  Copyright (C) 1998-2007   The R Development Core Team.
+ *  Copyright (C) 2008-2014  Andrew R. Runnalls.
+ *  Copyright (C) 2014 and onwards the Rho Project Authors.
+ *
+ *  Rho is not part of the R project, and bugs and other issues should
+ *  not be reported via r-bugs or other R project channels; instead refer
+ *  to the Rho website.
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, a copy is available at
+ *  https://www.R-project.org/Licenses/
+ */
+
+#include <assert.h>
+#include <stdio.h>
+
+#include <cstdint>
+#include <cstdlib>
+#include <functional>
+#include <limits>
+#include <map>
+
+#include "rho/AllocationTable.hpp"
+#include "rho/GCNodeAllocator.hpp"
+#include "rho/AddressSanitizer.hpp"
+
+// The low pointer bits are masked out in the hash function.
+#define LOW_BITS (19)
+#define MAX_COLLISIONS (6)
```

Comment:
When the table gets large, it's going to need to be really sparse to satisfy this.  With n elements in a hash table of size m, the expected number of collisions ~ n^2 / m, so we'd need m ~ n^2.

Reasonable options would be to either make the number of allowed collisions proportional to the size of the table, or just resize based on the load factor.